sa/sa.go

```diff
@@ -786,15 +786,34 @@ func addRevokedCertificate(ctx context.Context, tx db.Executor, req *sapb.Revoke
 		return errors.New("cannot add revoked certificate with shard index 0")
 	}
 
+	// If this serial already exists in the revokedCertificates table, then the
+	// certificate has already been revoked. The RA has special logic for that
+	// case, so return the specific error that the RA is looking for.
+	var row struct {
+		Count int64
+	}
+	err := tx.SelectOne(ctx, &row, "SELECT COUNT(*) as count FROM revokedCertificates WHERE serial = ?", req.Serial)
```

Comment:
I think this needs to be `SELECT ... FOR UPDATE`. Previously discussed a similar issue here: https://github.com/letsencrypt/boulder/pull/8030#discussion_r1975740980

Docs for `FOR UPDATE` here: https://dev.mysql.com/doc/refman/8.4/en/select.html. Note that it creates a dependency on InnoDB as the storage engine (or another storage engine that supports page or row locks).

Note that in the previous case, for leasing CRL shards, you took a different approach to avoid the `FOR UPDATE`. I haven't fully paged that approach back into memory yet, but maybe you recall how it works and whether it would be appropriate here?